expression/builtin_cast.go

```diff
@@ -1699,6 +1716,10 @@ func BuildCastFunction4Union(ctx sessionctx.Context, expr Expression, tp *types.
 
 // BuildCastFunction builds a CAST ScalarFunction from the Expression.
 func BuildCastFunction(ctx sessionctx.Context, expr Expression, tp *types.FieldType) (res Expression) {
+	if hasSpecialCast(ctx, expr, tp) {
```

Comment:
this check should be added in the case of `types.ETInt`, `types.ETReal` and `types.ETString`

Reply:
Do you means that check `tp.Type` here and call `hasSpecialCast` only when it's `int`, `real` or `string`?

Reply:
Oh, I missed the type check in function `hasSpecialCast()`. How about s/hasSpecialCast/skipCastDayname/?

Reply:
I'm afraid that there may be more functions who have a special cast process, then we can add them all here and it's not just for `DayName`.

Reply:
okay